### PR TITLE
Reduce to structured example with library functions.

### LINKED ISSUE #123: Add Single Body Device Model
As discussed in our last call, it makes a lot of sense to add the ability to model a single body device (probably in heave). This will help us encapsulate code that is shared between all models, and which is shared between similar classes, such as bodies in heave. 

A proposed case study for this model was [WaveBot](https://mhkdr.openei.org/submissions/151), but others candidates may also be viable.

### LINKED ISSUE #142: License location in source code files
From @H0R5E is #140:

As discussed in our call, not using the GPL notice in a file is probably incorrect application of the license. To quote [How to use GNU licenses for your own software](https://www.gnu.org/licenses/gpl-howto.en.html):
> * Put a license notice in each file.

I had a look at how MATLAB does this and the rules are as follows:

 * If there is no docstring, the copyright notice is placed at the bottom of the file
 * If a docstring exists, the copyright notice is placed after the docstring but, crucially, there is a blank line between the docstring and the copyright notice. E.g.:

```
classdef ComplexConjugate < WecOptTool.control.AbsControl
    % Provides complex congugate control
    %
    % Seeks optimal power absorption, see `Falnes2002`_ for technical 
    % details.
    % 
    % .. _Falnes2002:
    %     https://www.cambridge.org/core/books/ocean-waves-and-\
    %     oscillating-systems/8A3366809DE5C1F916FF87F36C55C459
    
    % Copyright 2020 National Technology & Engineering Solutions of Sandia, 
    % LLC (NTESS). Under the terms of Contract DE-NA0003525 with NTESS, the 
    % U.S. Government retains certain rights in this software.
    %
    % This file is part of WecOptTool.
    % 
    %     WecOptTool is free software: you can redistribute it and/or 
    %     modify it under the terms of the GNU General Public License as 
    %     published by the Free Software Foundation, either version 3 of 
    %     the License, or (at your option) any later version.
    % 
    %     WecOptTool is distributed in the hope that it will be useful,
    %     but WITHOUT ANY WARRANTY; without even the implied warranty of
    %     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    %     GNU General Public License for more details.
    % 
    %     You should have received a copy of the GNU General Public License
    %     along with WecOptTool.  If not, see 
    %     <https://www.gnu.org/licenses/>.
```

The blank line stops the help system picking up the copyright declaration and sphinx doesn't read it either. I suggest we follow a similar approach.

### PR DESCRIPTION
**Pull request by H0R5E**
So, I have modified the basic.m RM3 example to a barebones pair of function calls, supported by some library functions to help out. I haven't updated the optimisation example yet, as I have to finish for the day.

Let me know your thoughts.

Closes #142 
Closes #123 

TODO:

- [x] Finish WaveBot example
- [x] Update inline docs
- [x] Update web docs
- [x] Update README
- [x] Added the results of running the test suite (in pdf form)
- [x] All new files contain the GPL header
- [x] If `example.m` has been modified, the content / line numbers in 
  `docs\user\example.rst` are still valid or have been fixed
- [x] add unit tests

### CODE REVIEW

**Comment by H0R5E** [comment]
OK, I the optimisation example for RM3 is working in this structure now as well. I've removed all the base types classes and the types function (SeaState has been moved to the top level), the Blueprint and Device classes and recreated all of the functionality directly in the script. It's a little less neat but certainly more transparent.

**Comment by H0R5E** [comment]
OK, so I have restructured the toolbox into a single namespace (`WecOptTool`) which is more in keeping with this being a support library. As part of that I have broken out the `WecOptLib.utils` package into a package called `system` and a package called `math`, as this was mostly what the functions in there were doing. I have also moved the tests out of the toolbox and into the root directory. To do this I consolidated the `tests.data` package into the `SeaState` class. I also added the `regularWave` function to the `SeaState` class.

The main task left (beyond finishing the waveBot example) is to update the inline and web docs for this new structure as they are now completely out of whack. I will start working on this next.

File: examples/WaveBot/WaveBot_caseA.m [hunk 1]
```diff
@@ -53,6 +53,7 @@
 fp = wp/(2*pi);
 Tp = 1/fp; 
 SS = WecOptTool.SeaState.regularWave(w,[A,Tp]);
+% w = SS.getRegularFrequencies(0.3);
```
Comment (ryancoe) [inline examples/WaveBot/WaveBot_caseA.m:56]: I have an error I cannot track down where if we uncomment this line, the check within `getPSPhasePower` fails

https://github.com/SNL-WaterPower/WecOptTool/blob/54ad294acd9a421a4e58da05ae6519768d27cbe6/examples/WaveBot/simulateDevice.m#L402

Comment (ryancoe) [reply examples/WaveBot/WaveBot_caseA.m]: resolved w/ 7f437435c0eeb2d68f0e0e2851bb00585b62d83e

**Comment by ryancoe** [comment]
@H0R5E - I made some decent progress on this today, but really not as much as I might have hoped. Most of what I did shouldn't have broad consequences, but here's two things that may:

1. I updated `/examples/WaveBot/simulateDevice.m` to use MATLAB's argument validation framework and use name-value pairs to handle things like `Zmax`, `Fmax`, etc.

https://github.com/SNL-WaterPower/WecOptTool/blob/1705a256ec938ac62d274b11685a747d18f9a8e5/examples/%2BWaveBot/simulateDevice.m#L24

2. I didn't like how we had files/functions with the same names for each of the examples. Not sure what the best solution is, but I went ahead and used [Package Folder naming](https://www.mathworks.com/help/matlab/matlab_oop/scoping-classes-with-packages.html) (`+WaveBot`) to alleviate this (1705a256ec938ac62d274b11685a747d18f9a8e5). Feel free to revert this if you're not a fan.

**Comment by H0R5E** [comment]
@ryancoe, I have no issues with 1., seems sensible given we have sacrifised a lot of error checking that was provided by the framework.

Regarding 2., I don't think it's the right approach. I understand you want to make them unique, but these examples should never be added to the path, so the functions will only be accessible locally and therefore they don't need to have unique names. Secondly, having repeated names is *useful* as it provides a standardised approach to setting up the problem which we can document (i.e. "create two functions with this purpose...", etc), and it's easy for users to start from an existing example and then modify. Thirdly, it adds unnecessary complication, and removing complication is one of the drivers behind the approach in this PR. In my opinion, it's outside of the user domain to start adding packages - they can if they want, but I don't think it's something we should make it look like they *need* to do. Yeah, so, in conclusion, I think we should reverse this one.

**Comment by ryancoe** [comment]
@H0R5E - I thought more about point 2 over the weekend. Given how MATLAB searches the working directory at the top of the search path, I do think there's good reason to distinguish the names. We would not be forcing users to apply this approach in their own models, but just making the examples less prone to errors. For example, if you are in `examples/RM3` running that `basic.m`, and then you open and run `examples/WaveBot/WaveBot_caseA.m` without changing directories, it will try to call the RM3 functions. Its true that the `+dir` naming is meant for packages, but it is really just a search path tool (unless I'm missing something).

**Comment by H0R5E** [comment]
You've gotta love MATLAB! The alternative option would be to put them into
a private folder which would only allow them to be called by a non private
function in the same folder. That's maybe a less complex approach to
packages while being more robust.

**Comment by H0R5E** [comment]
For instance if someone copied and pasted the wavebot example without
changing the names you could get the same problem.

**Comment by ryancoe** [comment]
On further thought, let's go with the way you originally had it. I still think some other solution would be good, but I'd rather get this wrapped up first.

**Comment by H0R5E** [comment]
@ryancoe I tried to recreate the issue you described " if you are in examples/RM3 running that basic.m, and then you open and run examples/WaveBot/WaveBot_caseA.m without changing directories". I put `disp("WAVEBOT");` into `\examples\WaveBot\designDevice.m` and `\examples\WaveBot\simulateDevice.m` then (from the root directory) I did:

```
cd examples\RM3
run("..\WaveBot\WaveBot_caseA.m")
```

and the (abridged) output that was returned was:

```
WAVEBOT
Simulation 1 of 3
WAVEBOT
Simulation 2 of 3
WAVEBOT
Simulation 3 of 3
WAVEBOT

          AvgPow     |MaxPow|    PowTHD_dBc     MaxPos     MaxVel     MaxPTO
          _______    ________    __________    ________    _______    ______

    CC    -146.54     996.52        NaN         0.22899    0.45799    3711.8
    P     -43.098     86.196        NaN        0.051564    0.10313    835.82
    PS    -123.81     626.55        NaN         0.16071    0.31584      2002
```

So, it looks like the folder where the script is being run from is taking precedence. The important part is that these folders shouldn't be placed on the path.

**Comment by H0R5E** [comment]
Test results:

[test_results.pdf](https://github.com/SNL-WaterPower/WecOptTool/files/5036202/test_results.pdf)

**Comment by ryancoe** [comment]
I was running tests and figured I'd let you know that this is passing on macOS w/ 2020b Update 4
[test_results.pdf](https://github.com/SNL-WaterPower/WecOptTool/files/5038957/test_results.pdf)

**Pull request merged**